File: insight/backend/app/features/lite_cut/timeline_math.py
"""Pure LiteCut source-time and timeline-time mathematics."""

from __future__ import annotations

from typing import Any

from .visual_material import (
    VISUAL_CONTENT_FIT_VALUES,
    VISUAL_FREEZE_DEFAULT_SEC,
    VISUAL_FREEZE_MAX_SEC,
    VISUAL_FREEZE_MIN_SEC,
    VISUAL_SPEED_DEFAULT,
    VISUAL_SPEED_MAX,
    VISUAL_SPEED_MIN,
)

MIN_CLIP_VISIBLE_SEC = 0.1


def clip_duration_sec(clip: dict[str, Any]) -> float:
    trim_in = float(clip.get("trim_in") or 0)
    trim_out = clip.get("trim_out")
    if trim_out is not None:
        return max(MIN_CLIP_VISIBLE_SEC, float(trim_out) - trim_in)
    if clip.get("duration") is not None:
        return max(MIN_CLIP_VISIBLE_SEC, float(clip.get("duration") or 0) - trim_in)
    meta = clip.get("meta") if isinstance(clip.get("meta"), dict) else {}
    if meta.get("duration_sec") is not None:
        return max(MIN_CLIP_VISIBLE_SEC, float(meta["duration_sec"]) - trim_in)
    return 5.0


def clip_speed(clip: dict[str, Any]) -> float:
    try:
        speed = float(clip.get("speed") or VISUAL_SPEED_DEFAULT)
    except (TypeError, ValueError):
        speed = VISUAL_SPEED_DEFAULT
    return max(VISUAL_SPEED_MIN, min(VISUAL_SPEED_MAX, speed))


def clip_speed_keyframes(clip: dict[str, Any]) -> list[tuple[float, float]]:
    trim_in = max(0.0, float(clip.get("trim_in") or 0.0))
    trim_out = trim_in + clip_duration_sec(clip)
    points: list[tuple[float, float]] = []
    for raw in clip.get("speed_keyframes") or []:
        if not isinstance(raw, dict):
            continue
        try:
            source_sec = max(trim_in, min(trim_out, float(raw.get("source_sec"))))
            speed = max(VISUAL_SPEED_MIN, min(VISUAL_SPEED_MAX, float(raw.get("speed"))))
        except (TypeError, ValueError):
            continue
        points.append((source_sec, speed))
    points.sort(key=lambda point: point[0])
    deduplicated: list[tuple[float, float]] = []
    for point in points:
        if deduplicated and abs(deduplicated[-1][0] - point[0]) <= 1e-6:
            deduplicated[-1] = point
        else:
            deduplicated.append(point)
    if len(deduplicated) < 2:
        return []
    if deduplicated[0][0] > trim_in + 1e-6:
        deduplicated.insert(0, (trim_in, clip_speed(clip)))
    if deduplicated[-1][0] < trim_out - 1e-6:
        deduplicated.append((trim_out, deduplicated[-1][1]))
    return deduplicated


def clip_speed_segments(clip: dict[str, Any]) -> list[tuple[float, float, float]]:
    trim_in = max(0.0, float(clip.get("trim_in") or 0.0))
    trim_out = trim_in + clip_duration_sec(clip)
    points = clip_speed_keyframes(clip)
    if not points:
        return [(trim_in, trim_out, clip_speed(clip))]
    return [(left_t, right_t, speed) for (left_t, speed), (right_t, _) in zip(points[:-1], points[1:]) if right_t - left_t > 1e-6]


def clip_has_speed_ramp(clip: dict[str, Any]) -> bool:
    return len(clip_speed_keyframes(clip)) >= 2


def clip_timeline_time_for_source(clip: dict[str, Any], source_sec: float) -> float:
    trim_in = max(0.0, float(clip.get("trim_in") or 0.0))
    source = max(trim_in, min(trim_in + clip_duration_sec(clip), float(source_sec)))
    timeline = 0.0
    for start, end, speed in clip_speed_segments(clip):
        if source <= start:
            break
        timeline += (min(source, end) - start) / speed
        if source <= end:
            break
    return timeline


def clip_media_timeline_duration_sec(clip: dict[str, Any]) -> float:
    trim_in = max(0.0, float(clip.get("trim_in") or 0.0))
    return max(MIN_CLIP_VISIBLE_SEC, clip_timeline_time_for_source(clip, trim_in + clip_duration_sec(clip)))


def clip_source_time_for_timeline(clip: dict[str, Any], timeline_sec: float) -> float:
    target = max(0.0, min(clip_media_timeline_duration_sec(clip), float(timeline_sec)))
    elapsed = 0.0
    for start, end, speed in clip_speed_segments(clip):
        timeline_length = (end - start) / speed
        if target <= elapsed + timeline_length + 1e-6:
            return start + max(0.0, target - elapsed) * speed
        elapsed += timeline_length
    return max(0.0, float(clip.get("trim_in") or 0.0)) + clip_duration_sec(clip)


def clip_reverse(clip: dict[str, Any]) -> bool:
    return bool(clip.get("reverse"))


def clip_freeze_frame_sec(clip: dict[str, Any]) -> float:
    try:
        freeze = float(clip.get("freeze_frame_sec") or VISUAL_FREEZE_DEFAULT_SEC)
    except (TypeError, ValueError):
        freeze = VISUAL_FREEZE_DEFAULT_SEC
    return max(VISUAL_FREEZE_MIN_SEC, min(VISUAL_FREEZE_MAX_SEC, freeze))


def clip_preserve_pitch(clip: dict[str, Any]) -> bool:
    return clip.get("preserve_pitch") is not False


def clip_canvas_fit(clip: dict[str, Any], fallback: str = "contain") -> str:
    raw = str(clip.get("content_fit") or "").strip().lower()
    if raw in VISUAL_CONTENT_FIT_VALUES and raw != "inherit":
        return raw
    fit = str(fallback or "contain").strip().lower()
    return fit if fit in VISUAL_CONTENT_FIT_VALUES and fit != "inherit" else "contain"


def clip_timeline_duration_sec(clip: dict[str, Any]) -> float:
    return clip_media_timeline_duration_sec(clip) + clip_freeze_frame_sec(clip)

File: insight/frontend/src/features/lite-cut/domain/timelineMath.js
import {
  VISUAL_FREEZE_DEFAULT_SEC,
  VISUAL_FREEZE_MAX_SEC,
  VISUAL_FREEZE_MIN_SEC,
  VISUAL_SPEED_DEFAULT,
  VISUAL_SPEED_MAX,
  VISUAL_SPEED_MIN,
  visualContentFit,
} from "./visualMaterial.js";

export const MIN_CLIP_VISIBLE_SEC = 0.1;

export function clipPlaybackSpeed(clip) {
  const speed = Number(clip?.speed);
  return Number.isFinite(speed) && speed > 0 ? Math.max(VISUAL_SPEED_MIN, Math.min(VISUAL_SPEED_MAX, speed)) : VISUAL_SPEED_DEFAULT;
}

export function clipTrimmedSourceDuration(clip) {
  if (!clip) return 5;
  const trimOut = clip.trim_out;
  const trimIn = Number(clip.trim_in) || 0;
  if (trimOut != null && Number.isFinite(Number(trimOut))) {
    return Math.max(MIN_CLIP_VISIBLE_SEC, Number(trimOut) - trimIn);
  }
  const meta = clip.meta;
  if (meta && Number.isFinite(Number(meta.duration_sec))) {
    return Math.max(MIN_CLIP_VISIBLE_SEC, Number(meta.duration_sec) - trimIn);
  }
  return 5;
}

export function normalizedClipSpeedKeyframes(clip) {
  const trimIn = Math.max(0, Number(clip?.trim_in) || 0);
  const trimOut = trimIn + clipTrimmedSourceDuration(clip);
  const fallback = clipPlaybackSpeed(clip);
  const points = [];
  for (const raw of clip?.speed_keyframes || []) {
    if (!raw || typeof raw !== "object") continue;
    const sourceSec = Number(raw.source_sec);
    const speed = Number(raw.speed);
    if (!Number.isFinite(sourceSec) || !Number.isFinite(speed)) continue;
    points.push({
      source_sec: Math.max(trimIn, Math.min(trimOut, sourceSec)),
      speed: Math.max(VISUAL_SPEED_MIN, Math.min(VISUAL_SPEED_MAX, speed)),
    });
  }
  points.sort((a, b) => a.source_sec - b.source_sec);
  const deduped = [];
  for (const point of points) {
    const index = deduped.findIndex((item) => Math.abs(item.source_sec - point.source_sec) < 0.0001);
    if (index >= 0) deduped[index] = point;
    else deduped.push(point);
  }
  if (deduped.length < 2) return [];
  if (deduped[0].source_sec > trimIn + 0.0001) deduped.unshift({ source_sec: trimIn, speed: fallback });
  if (deduped.at(-1).source_sec < trimOut - 0.0001) deduped.push({ source_sec: trimOut, speed: deduped.at(-1).speed });
  return deduped;
}

export function clipSpeedSegments(clip) {
  const trimIn = Math.max(0, Number(clip?.trim_in) || 0);
  const trimOut = trimIn + clipTrimmedSourceDuration(clip);
  const points = normalizedClipSpeedKeyframes(clip);
  if (!points.length) return [{ sourceStart: trimIn, sourceEnd: trimOut, speed: clipPlaybackSpeed(clip) }];
  return points.slice(0, -1).map((point, index) => ({
    sourceStart: point.source_sec,
    sourceEnd: points[index + 1].source_sec,
    speed: point.speed,
  })).filter((segment) => segment.sourceEnd - segment.sourceStart > 0.0001);
}

export function clipTimelineTimeForSource(clip, sourceSec) {
  const trimIn = Math.max(0, Number(clip?.trim_in) || 0);
  const source = Math.max(trimIn, Math.min(trimIn + clipTrimmedSourceDuration(clip), Number(sourceSec) || trimIn));
  let timeline = 0;
  for (const segment of clipSpeedSegments(clip)) {
    if (source <= segment.sourceStart) break;
    timeline += (Math.min(source, segment.sourceEnd) - segment.sourceStart) / segment.speed;
    if (source <= segment.sourceEnd) break;
  }
  return timeline;
}

export function clipMediaTimelineDuration(clip) {
  return Math.max(MIN_CLIP_VISIBLE_SEC, clipTimelineTimeForSource(clip, (Number(clip?.trim_in) || 0) + clipTrimmedSourceDuration(clip)));
}

export function clipSourceTimeForTimeline(clip, timelineSec) {
  const target = Math.max(0, Math.min(clipMediaTimelineDuration(clip), Number(timelineSec) || 0));
  let elapsed = 0;
  for (const segment of clipSpeedSegments(clip)) {
    const timelineLength = (segment.sourceEnd - segment.sourceStart) / segment.speed;
    if (target <= elapsed + timelineLength + 0.000001) {
      return segment.sourceStart + Math.max(0, target - elapsed) * segment.speed;
    }
    elapsed += timelineLength;
  }
  return Math.max(0, Number(clip?.trim_in) || 0) + clipTrimmedSourceDuration(clip);
}

export function clipSpeedAtTimeline(clip, timelineSec) {
  const source = clipSourceTimeForTimeline(clip, timelineSec);
  const segment = clipSpeedSegments(clip).find((item) => source >= item.sourceStart - 0.0001 && source <= item.sourceEnd + 0.0001);
  return segment?.speed ?? clipPlaybackSpeed(clip);
}

export function clipFreezeFrameSec(clip) {
  const freeze = Number(clip?.freeze_frame_sec);
  return Number.isFinite(freeze) ? Math.max(VISUAL_FREEZE_MIN_SEC, Math.min(VISUAL_FREEZE_MAX_SEC, freeze)) : VISUAL_FREEZE_DEFAULT_SEC;
}

export function clipReversePlayback(clip) {
  return Boolean(clip?.reverse);
}

export function clipPreservePitch(clip) {
  return clip?.preserve_pitch !== false;
}

export function clipCanvasFit(clip, fallback = "contain") {
  return visualContentFit(clip, fallback);
}

export function clipTimelineDuration(clip) {
  return clipMediaTimelineDuration(clip) + clipFreezeFrameSec(clip);
}

export function clipSourceDuration(clip) {
  return clipTimelineDuration(clip);
}

export function ensureClipSourceDuration(clip) {
  if (!clip) return 5;
  if (!clip.meta || typeof clip.meta !== "object") clip.meta = {};
  const existing = Number(clip.meta.duration_sec);
  if (existing > 0) return existing;
  const trimIn = Number(clip.trim_in) || 0;
  const trimOut = Number(clip.trim_out);
  const inferred = trimOut > trimIn ? trimOut : trimIn + clipTrimmedSourceDuration(clip);
  clip.meta.duration_sec = Math.max(MIN_CLIP_VISIBLE_SEC, inferred);
  return clip.meta.duration_sec;
}

export function clipSourceMediaDuration(clip) {
  if (!clip) return 5;
  const meta = clip.meta;
  if (meta && Number.isFinite(Number(meta.duration_sec)) && Number(meta.duration_sec) > 0) {
    return Number(meta.duration_sec);
  }
  return ensureClipSourceDuration(clip);
}

export function clipMaxTimelineEnd(clip) {
  const start = Number(clip.timeline_start) || 0;
  const sourceDur = clipSourceMediaDuration(clip);
  const extended = { ...clip, trim_out: sourceDur };
  return start + Math.max(MIN_CLIP_VISIBLE_SEC, clipTimelineTimeForSource(extended, sourceDur)) + clipFreezeFrameSec(clip);
}

export function clipMaxTimelineStartForLeftTrim(clip) {
  const start = Number(clip.timeline_start) || 0;
  const end = clipTimelineEnd(clip);
  const sourceDur = clipSourceMediaDuration(clip);
  const extended = { ...clip, trim_out: sourceDur };
  const maxFromSource = start + Math.max(0, clipTimelineTimeForSource(extended, sourceDur - MIN_CLIP_VISIBLE_SEC));
  return Math.min(end - MIN_CLIP_VISIBLE_SEC, maxFromSource);
}

export function clipTimelineEnd(clip) {
  return (Number(clip.timeline_start) || 0) + clipSourceDuration(clip);
}
